benchmarks/misc/end2end_cases/rasterizer.py

```diff
@@ -0,0 +1,159 @@
+                if self.bbox_intersect(tile_min, tile_max, tri_min, tri_max):
+                    self.block_indicies[i, j, idx] = t
+                    idx = idx + 1
+            self.block_num_triangles[i, j] = idx
+
+    @ti.kernel
+    def rasterize(self):
+        for i, j in self.pixels:
+            for k in range(self.block_num_triangles[i // self.tile_size,
+                                                    j // self.tile_size]):
+                idx = self.block_indicies[i // self.tile_size,
+                                          j // self.tile_size, k]
+                A, B, C = self.A[idx], self.B[idx], self.C[idx]
+                c0, c1, c2 = self.c0[idx], self.c1[idx], self.c2[idx]
+
+                for subi, subj in ti.ndrange(self.num_spp_sqrt,
+                                             self.num_spp_sqrt):
+                    P = ti.Vector([
+                        i + (subi + 0.5) / self.num_spp_sqrt,
+                        j + (subj + 0.5) / self.num_spp_sqrt
+                    ])
+                    result, alpha, beta, gamma = self.point_in_triangle(
+                        P, A, B, C)
+
+                    if result:
+                        interpolated_color = c0 * alpha + c1 * beta + c2 * gamma
+                        self.samples[i, j, subi, subj] = interpolated_color
+
+            samples_sum = ti.Vector([0.0, 0.0, 0.0])
+            for subi, subj in ti.ndrange(self.num_spp_sqrt, self.num_spp_sqrt):
+                samples_sum += self.samples[i, j, subi, subj]
+            self.pixels[i, j] = samples_sum / self.num_samples_per_pixel
+
+
+def e2e_rasterizer(test_arch):
+    basic_repeat_times = 100
+    arch_repeat_times = 10 if test_arch == ti.cuda else 1
+
+    tile_size = 8  # Size of a tile
+    width, height = 800, 600  # Size of framebuffer
+    num_samples_per_pixel = 4  # Number of samples per pixel
+    num_spp_sqrt = int(math.sqrt(num_samples_per_pixel))
+    num_triangles = 256  # Number of random colored triangles to be generated
+
+    ti.init(kernel_profiler=True, arch=test_arch)
+
+    print('    initializing ...')
+    triangles = TriangleRasterizer(num_triangles, tile_size, width, height,
+                                   num_samples_per_pixel, num_spp_sqrt)
+    for i in range(num_triangles):
+        triangles.set_triangle(i % num_triangles,
+                               ti.Vector(np.random.rand(2) * [width, height]),
+                               ti.Vector(np.random.rand(2) * [width, height]),
+                               ti.Vector(np.random.rand(2) * [width, height]),
+                               ti.Vector(np.random.rand(3)),
+                               ti.Vector(np.random.rand(3)),
+                               ti.Vector(np.random.rand(3)))
+
+    triangles.samples.fill(ti.Vector([1.0, 1.0, 1.0]))
+    triangles.pixels.fill(ti.Vector([1.0, 1.0, 1.0]))
+    triangles.tile_culling()
+    triangles.rasterize()
+
+    print('    profiling begin ...')
```

Comment:
FYI: The code is basically copied from examples/, and the profiling section is at the end of the file